airflow-core/src/airflow/jobs/asyncio_monitor.py

```diff
@@ -0,0 +1,244 @@
+        # Reschedule next beat
+        self.loop.call_later(self.heartbeat_interval, self._heartbeat)
+
+    def _watchdog(self) -> None:
+        while self._running:
+            time.sleep(self.heartbeat_interval)
+            now = perf_counter()
+            gap = now - self._hb_perf
+
+            if gap > self.threshold:
+                # We are in a stall
+                if self._incident is None:
+                    # Stall START
+                    self._incident = StallIncident(
+                        started_at_perf=now - gap,  # best estimate: when heartbeat stopped advancing
+                        ended_at_perf=None,
+                        threshold=self.threshold,
+                        samples=[],
+                        suspected_tasks=[],
+                        started_at_utc=_utc_now_str(),
+                        ended_at_utc="",
+                    )
+                    self._last_report_perf = 0.0
+                    self._sample_and_log(now, phase="start")
+
+                # During stall: periodic updates
+                elif now - self._last_report_perf >= self.min_report_interval:
+                    self._sample_and_log(now, phase="update")
+
+            else:
+                # No stall *or* stall ended: close & correlate if needed
+                if self._incident is not None:
+                    self._incident.ended_at_perf = now
+                    self._incident.ended_at_utc = _utc_now_str()
+                    incident = self._incident
+                    self._incident = None
+                    self._post_incident_correlation(incident)
+
+    def _sample_and_log(self, now_perf: float, *, phase: str) -> None:
+        stack_text = self._capture_loop_stack_bounded()
+        if stack_text is None:
+            stack_text = "<loop thread stack unavailable>"
+
+        sample = StallSample(taken_at_perf=now_perf, stack_text=stack_text)
+        if self._incident:
+            self._incident.samples.append(sample)
+
+        if phase == "start":
+            log.warning("Event loop stall detected (gap≥%.3fs). Captured loop stack.", self.threshold)
+        else:
+            log.warning(
+                "Event loop still stalled (%.3fs since start).",
+                now_perf - (self._incident.started_at_perf if self._incident else now_perf),
+            )
+        log.warning("%s", stack_text)
+        Stats.incr("triggers.blocked_main_thread")
+
+        self._last_report_perf = now_perf
+
+    def _capture_loop_stack_bounded(self) -> str | None:
+        ident = self._loop_thread_ident
+        if ident is None:
+            return None
+        try:
+            frame = sys._current_frames().get(ident)
+            if not frame:
+                return None
+            return _format_stack_from_frame(frame, self.max_frames)
+        except Exception:
+            return None
+
+    def _post_incident_correlation(self, incident: StallIncident) -> None:
```

Comment:
prob rename since not really correlating anymore?  i think it previously meant correlate _to task_ but now, to keep it simple, we don't bother with that